File: go/leet_code/165.go
// Package leet_code provides solutions for LeetCode problems.
package leet_code

import (
	"strconv"
	"strings"
)

// compareVersion compares two version strings and returns an integer according to the comparison result.
// It splits the version strings by ".", converts each part to an integer, and compares them part by part.
// If version1 is greater than version2, it returns 1.
// If version1 is less than version2, it returns -1.
// If version1 is equal to version2, it returns 0.
//
// Parameters:
// version1: A string representing the first version to compare. It should be a non-empty string consisting of digits and periods.
// version2: A string representing the second version to compare. It should be a non-empty string consisting of digits and periods.
//
// Returns:
// An integer indicating the comparison result. It returns 1 if version1 > version2, -1 if version1 < version2, and 0 if version1 = version2.
func compareVersion(version1 string, version2 string) int {
	// Split the version strings by "."
	rev1 := strings.Split(version1, ".")
	rev2 := strings.Split(version2, ".")

	// Convert each part of the version strings to an integer
	revInt1 := make([]int, len(rev1))
	revInt2 := make([]int, len(rev2))

	for i := 0; i < len(rev1); i++ {
		revInt1[i], _ = strconv.Atoi(rev1[i])
	}
	for i := 0; i < len(rev2); i++ {
		revInt2[i], _ = strconv.Atoi(rev2[i])
	}

	// Compare the integer parts of the version strings
	i := 0
	j := 0

	for i < len(revInt1) && j < len(revInt2) {
		if revInt1[i] < revInt2[j] {
			return -1
		} else if revInt1[i] > revInt2[j] {
			return 1
		}
		i++
		j++
	}

	// Check the remaining parts of the version strings
	for i < len(revInt1) {
		if revInt1[i] > 0 {
			return 1
		}
		i++
	}
	for j < len(revInt2) {
		if revInt2[j] > 0 {
			return -1
		}
		j++
	}

	// If all parts are equal, return 0
	return 0
}

File: javascript/leet_code/165.mjs
/**
 * @fileoverview This module provides a function to compare two version strings.
 * It is used to determine if one version is greater than, equal to, or less than the other.
 * The version strings are expected to be in the format 'x.y.z' where x, y, and z are integers.
 * The comparison is done from left to right, meaning it first compares the major version (x),
 * then the minor version (y), and finally the patch version (z).
 */

/**
 * Compares two version strings.
 *
 * @param {string} ver1 - The first version string to compare. It should be in the format 'x.y.z'.
 * @param {string} ver2 - The second version string to compare. It should be in the format 'x.y.z'.
 * @returns {number} - Returns 0 if the versions are equal, -1 if ver1 is less than ver2, and 1 if ver1 is greater than
 *     ver2.
 */
const compareVersion = function (ver1, ver2) {
    // Split the version strings into arrays of integers
    const nums1 = ver1.split('.').map(n => parseInt(n));
    const nums2 = ver2.split('.').map(n => parseInt(n));

    // Compare the corresponding elements in the arrays
    while (nums1.length > 0 && nums2.length > 0) {
        const num1 = nums1.shift();
        const num2 = nums2.shift();

        // If the elements are equal, continue to the next pair
        if (num1 === num2) {
            continue;
        }

        // If the element from ver1 is less than the element from ver2, return -1
        if (num1 < num2) {
            return -1;
        } else {
            // If the element from ver1 is greater than the element from ver2, return 1
            return 1;
        }
    }

    // If there are remaining elements in ver1 and they are not zero, return 1
    while (nums1.length > 0) {
        const num = nums1.shift();
        if (num === 0) {
            continue;
        }
        return 1;
    }

    // If there are remaining elements in ver2 and they are not zero, return -1
    while (nums2.length > 0) {
        const num = nums2.shift();
        if (num === 0) {
            continue;
        }
        return -1;
    }

    // If all elements have been compared and are equal, return 0
    return 0;
};
